shopfloor_reception/services/reception.py

```diff
@@ -0,0 +1,877 @@
+        picking = self.env["stock.picking"].browse(picking_id)
+        if all(line.qty_done == 0 for line in picking.move_line_ids):
+            # If no line has been processed, refuse to set the picking as done
+            return self._response_for_select_line(
+                picking, message=self.msg_store.transfer_no_qty_done()
+            )
+        if not confirmation:
+            to_backorder = picking._check_backorder()
+            if to_backorder:
+                # Not all lines are fully done, ask the user to confirm the
+                # backorder creation
+                return self._response_for_select_line(
+                    picking, message=self.msg_store.transfer_confirm_done()
+                )
+            # all lines are done, ask the user to confirm anyway
+            return self._response_for_select_line(
+                picking, message=self.msg_store.need_confirmation()
+            )
+        backorders_before = picking.backorder_ids
+        picking._action_done()
+        backorders_after = picking.backorder_ids - backorders_before
+        # Remove user_id on the backorder, if any
+        backorders_after.user_id = False
+        return self._response_for_select_document(
+            message=self.msg_store.transfer_done_success(picking)
+        )
+
+    def set_lot(
+        self, picking_id, selected_line_ids, lot_name=None, expiration_date=None
+    ):
+        """Set lot and its expiration date
+
+        Input:
+            barcode: The barcode of a lot
+            expiration_date: The expiration_date
+
+        transitions:
+          - select_line: User clicked on back
+          - set_lot: Barcode not found. Ask user to create one from barcode
+          - set_lot: expiration_date has been set on the selected line
+          - set_lot: lot_it has been set on the selected line
+          - set_lot: Error: expiration_date is required
+          - set_quantity: User clicked on the confirm button
+        """
+        picking = self.env["stock.picking"].browse(picking_id)
+        selected_lines = self.env["stock.move.line"].browse(selected_line_ids)
+        lot_model = self.env["stock.production.lot"]
+        search = self._actions_for("search")
+        if lot_name:
+            for line in selected_lines:
+                lot = search.lot_from_scan(
+                    lot_name, products=selected_lines.mapped("product_id")
+                )
+                if not lot:
+                    lot = lot_model.create(
+                        {
+                            "name": lot_name,
+                            "product_id": line.product_id.id,
+                            "company_id": self.env.company.id,
+                            "use_expiration_date": line.product_id.use_expiration_date,
+                        }
+                    )
+                line.lot_id = lot.id
+                # TODO: Dunno why, I have to explicitely call the onchange method
```

Comment:
Not sure to understand this comment?
It's not a good thing to call the onchange methods here anyway? Looking at the code of `product_expiry`, this onchange updates the `expiration_date` field of the line, that seems important.

Reply:
Yeah this is important. The thing is that changing the lot doesn't update the expiration date.
We have to force it, despite what suggests this piece of code https://github.com/odoo/odoo/blob/14.0/addons/product_expiry/models/stock_move_line.py#L29-L36

Reply:
Onchange methods are a UI feature, they are never called when fields are updated on server-side, so it's normal to call them manually.